saleor/webhook/observability/tests/test_worker.py

```diff
@@ -0,0 +1,151 @@
+import time
+from datetime import timedelta
+from functools import partial
+from unittest.mock import patch
+
+import pytest
+from redis.exceptions import RedisError
+
+from .. import worker
+from .conftest import KEY, MAX_SIZE
+
+BATCH_SIZE = 10
+REPORT_PERIOD = timedelta(seconds=0.01)
+timeout = REPORT_PERIOD.total_seconds()
+safe_delay = timeout * 2
+
+
+@pytest.fixture
+def init_worker(settings):
+    settings.OBSERVABILITY_BUFFER_BATCH_SIZE = BATCH_SIZE
+    settings.OBSERVABILITY_REPORT_PERIOD = REPORT_PERIOD
+    worker.init()
+    yield
+    worker.shutdown_worker()
+
+
+def test_init(init_worker):
+    assert isinstance(worker._worker, worker.BackgroundWorker)
+
+
+def test_put_event_when_worker_not_initialized():
+    worker.shutdown_worker()
+    assert worker.put_event("buffer", "payload") is False
+
+
+@patch("saleor.webhook.observability.worker.get_buffer")
+def test_buffer_put_multi_key_events(mock_get_buffer, buffer):
+    mock_get_buffer.return_value = buffer
+    events = {KEY: ["payload"], "buffer_b": ["payload"]}
+    assert worker.buffer_put_multi_key_events(events) is True
+
+
+@patch("saleor.webhook.observability.worker.get_buffer")
+def test_buffer_put_multi_key_events_when_buffer_is_full(mock_get_buffer, buffer):
+    mock_get_buffer.return_value = buffer
+    for i in range(MAX_SIZE):
+        buffer.put_event("payload")
+    assert buffer.size() == MAX_SIZE
+    events = {KEY: ["payload"], "buffer_b": ["payload"]}
+    assert worker.buffer_put_multi_key_events(events) is False
+
+
+@patch("saleor.webhook.observability.worker.buffer_put_multi_key_events")
+def test_background_worker(mock_buffer_put_multi_key_events, init_worker):
+    buffer_name, payload = "buffer_a", "payload-{}"
+    for i in range(BATCH_SIZE + 1):
+        worker.put_event(buffer_name, partial(payload.format, i))
+    mock_buffer_put_multi_key_events.assert_called_once_with(
+        {buffer_name: [payload.format(i) for i in range(BATCH_SIZE)]}
+    )
+
+
+@patch("saleor.webhook.observability.worker.buffer_put_multi_key_events")
+def test_background_worker_put_to_buffer_exception(
+    mock_buffer_put_multi_key_events, init_worker
+):
+    mock_buffer_put_multi_key_events.side_effect = [RedisError, True]
+    for i in range(BATCH_SIZE * 2):
+        worker.put_event("buffer_a", lambda: "payload")
+    time.sleep(safe_delay)
```

Comment:
IMHO I don't like using `time.sleep` in the logic/tests. There is always a risk, that the sleep will not be enough or something. I think that we need to consider better solution for testing it.

Reply:
Agree, working on it 👍